url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/workflow/WorkflowEngine.java

```diff
@@ -0,0 +1,506 @@
+				}
+			}
+		}
+		
+		moveData(task);		
+		dispatchTasks();
+	}
+	
+	private void processCloudletCancel(SimEvent ev) {
+		Cloudlet cloudlet = (Cloudlet) ev.getData();
+		
+		if (cloudlet!=null){
+			getCloudletReceivedList().add(cloudlet);
+			Log.printLine(CloudSim.clock()+": Task "+cloudlet.getCloudletId()+" cancelled.");
+				
+			Task task = cloudletTaskMap.remove(cloudlet);
+			freeVmList.put(task.getVmId(), true);
+		}
+		
+		dispatchTasks();
+	}
+	
+	private void processDataItemAvailable(SimEvent ev) {
+		Transmission tr = (Transmission) ev.getData();
+		Log.printLine(CloudSim.clock()+": DataItem #"+tr.getDataItem().getId()+" is now available at VM #"+tr.getDestinationId());
+		removeOngoingTransmission(tr.getSourceId(),tr.getDestinationId(),tr.getDataItem().getId());
+		
+		dispatchTasks();
+	}
+	
+	private void processStartDelayEvent() {
+		// we gave data center enough time to initialize. Start the action...
+		datacenterId = CloudSim.getCloudResourceList().get(0);
+		sendNow(datacenterId, CloudSimTags.RESOURCE_CHARACTERISTICS, getId());
+	}
+
+	@Override
+	public void shutdownEntity() {
+		//do nothing
+	}
+
+	@Override
+	public void startEntity() {
+		//give time to the data center to start
+		send(getId(),2,DELAY_START_EVENT);
+	}
+	
+	//Triggers the process of copying data necessary in other vms
+	private void moveData(Task task) {
+		int originVm = task.getVmId();
+		
+		for(DataItem data:task.getOutput()){
+			for (int destVm: dataRequiredLocation.get(data.getId())){
+				WETransmission tr = new WETransmission(task, originVm, destVm,data.getId());
+							
+				if (!data.isAvailableAt(destVm)){//if origin != destiny and data is not in destiny, trigger transfer
+					Log.printLine(CloudSim.clock()+": Transferring dataItem #"+data.getId()+" from VM #"+originVm+" to VM #"+destVm);
+					
+					pendingTransmissions.add(tr);
+					if (runningVmList.containsKey(destVm) && runningVmList.get(destVm)==true){
+						//if vm was not created, delay transmission
+						TransferDataEvent event = new TransferDataEvent(data,originVm,destVm);
+						sendNow(datacenterId,WorkflowDatacenter.TRANSFER_DATA_ITEM,event);
+					}
+				} else {
+					//Log.printLine(CloudSim.clock()+": DataItem #"+data.getId()+" is already available at VM #"+destVm);
+				}
+			}
+		}
+	}
+	
+	//Triggers the process of copying data necessary to a delayed vm
+	private void moveData(Task task, int vmId) {
+		int originVm = task.getVmId();
+		
+		for(DataItem data:task.getOutput()){
+			for (int destVm: dataRequiredLocation.get(data.getId())){
+				if (destVm==vmId && !data.isAvailableAt(destVm)){
+					Log.printLine(CloudSim.clock()+": Transferring dataItem #"+data.getId()+" from VM #"+originVm+" to VM #"+destVm);
+					TransferDataEvent event = new TransferDataEvent(data,originVm,destVm);
+					sendNow(datacenterId,WorkflowDatacenter.TRANSFER_DATA_ITEM,event);
+				}
+			}
+		}
+	}
+	
+	private void dispatchTasks() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 291-297 into a separate method.